### PR TITLE
JBPM-5862-2: Stunner - Diagram tree explorer improvements.

### PR DESCRIPTION
**Pull request by alepintus**
Refactored code to improve the performance of the uberfire tree and stunner tree explorer.
For more information see [this ticket](https://issues.jboss.org/browse/JBPM-5862).

This PR should be merged with [this kie-wb-common PR](https://github.com/kiegroup/kie-wb-common/pull/918) 

@romartin can you please review this PR?

Thanks.

### CODE REVIEW

**Comment by hasys** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
Jenkins please retest this

**Comment by romartin** [comment]
Jenkins execute full downstream build

**Comment by romartin** [comment]
Jenkins execute full downstream build

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/TreeItem.java [hunk 1]
```diff
@@ -370,6 +384,19 @@ FlowPanel getContent() {
         return content;
     }
 
+    @Override
+    public boolean equals(final Object other) {
```
Comment (romartin) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/TreeItem.java]: Missing to override the `hashCode` method too.

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java [hunk 2]
```diff
@@ -94,6 +94,53 @@ public T addItem(final T item) {
         return item;
     }
 
+    @SuppressWarnings("unchecked")
+    public T getItemByUuid(String uuid) {
+
```
Comment (romartin) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java]: just minor thing - please remove all these empty lines that appear on these classe, there are lots :).

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java [hunk 3]
```diff
@@ -94,6 +94,53 @@ public T addItem(final T item) {
...
+        T item[] = (T[]) new TreeItem[1];
```
Comment (romartin) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java]: can be `final`

File: uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java [hunk 4]
```diff
@@ -94,6 +94,53 @@ public T addItem(final T item) {
...
+
+        getItems().forEach(i -> {
```
Comment (romartin) [inline uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java]: I think the logic for both `getItemByUuid` can be improved and make it better readable in general. For example, I would try to move all logic for looking by uuid into a TreeItem into a single method, it checks iteself and its children, and in both cases you can avoid the iterations, conditional expressions etc by creating an `infinite Stream` (considering reasonable tree sizes) and using the regular `Stream API` methods.
NOTE: You can create the `Stream` instance like this `StreamSupport.stream(getItems().spliterator(), false)`
WDYT?

**Comment by romartin** [comment]
Please @alepintus , rebase this commit as well, let's see if the downstream build finally works.

**Comment by romartin** [comment]
Jenkins please retest this.

**Pull request merged**